libraries/AP_DAL/AP_DAL.h

```diff
@@ -0,0 +1,363 @@
+#pragma once
+
+#include <stdint.h>
+
+#include <cstddef>
+
+#include "AP_DAL_InertialSensor.h"
+#include "AP_DAL_Baro.h"
+#include "AP_DAL_GPS.h"
+#include "AP_DAL_RangeFinder.h"
+#include "AP_DAL_Compass.h"
+#include "AP_DAL_Airspeed.h"
+#include "AP_DAL_Beacon.h"
+#include "AP_DAL_VisualOdom.h"
+
+#include "LogStructure.h"
+
+#define DAL_CORE(c) AP::dal().logging_core(c)
+
+#if APM_BUILD_TYPE(APM_BUILD_Replay)
+class NavEKF2;
+class NavEKF3;
+#endif
+
+class AP_DAL {
+public:
+
+    enum class FrameType {
+        InitialiseFilterEKF2 = 1<<0,
+        UpdateFilterEKF2 = 1<<1,
+        InitialiseFilterEKF3 = 1<<2,
+        UpdateFilterEKF3 = 1<<3,
+        LogWriteEKF2 = 1<<4,
+        LogWriteEKF3 = 1<<5,
+    };
+
+    enum class Event2 {
+        resetGyroBias             =  0,
+        resetHeightDatum          =  1,
+        setInhibitGPS             =  2,
+        setTakeoffExpected        =  3,
+        unsetTakeoffExpected      =  4,
+        setTouchdownExpected      =  5,
+        unsetTouchdownExpected    =  6,
+        setInhibitGpsVertVelUse   =  7,
+        unsetInhibitGpsVertVelUse =  8,
+        setTerrainHgtStable       =  9,
+        unsetTerrainHgtStable     = 10,
+        requestYawReset           = 11,
+        checkLaneSwitch           = 12,
+    };
+
+    enum class Event3 {
+        resetGyroBias             =  0,
+        resetHeightDatum          =  1,
+        setInhibitGPS             =  2,
+        setTakeoffExpected        =  3,
+        unsetTakeoffExpected      =  4,
+        setTouchdownExpected      =  5,
+        unsetTouchdownExpected    =  6,
+        setInhibitGpsVertVelUse   =  7,
+        unsetInhibitGpsVertVelUse =  8,
+        setTerrainHgtStable       =  9,
+        unsetTerrainHgtStable     = 10,
+        requestYawReset           = 11,
+        checkLaneSwitch           = 12,
+    };
+
+    // must remain the same as AP_AHRS_VehicleClass numbers-wise
+    enum class VehicleClass : uint8_t {
+        UNKNOWN,
+        GROUND,
+        COPTER,
+        FIXED_WING,
+        SUBMARINE,
+    };
+
+    AP_DAL() {}
+
+    static AP_DAL *get_singleton() {
+        if (!_singleton) {
+            _singleton = new AP_DAL();
+        }
+        return _singleton;
+    }
+
+    void start_frame(FrameType frametype);
+    void end_frame(void);
+    uint64_t micros64() const { return _RFRH.time_us; }
+    uint32_t micros() const { return _micros; }
+    uint32_t millis() const { return _millis; }
+
+    void log_event2(Event2 event);
+    void log_SetOriginLLH2(const Location &loc);
+    void log_writeDefaultAirSpeed2(float aspeed);
+
+    void log_event3(Event3 event);
+    void log_SetOriginLLH3(const Location &loc);
+    void log_writeDefaultAirSpeed3(float aspeed);
+    void log_writeEulerYawAngle(float yawAngle, float yawAngleErr, uint32_t timeStamp_ms, uint8_t type);
+
+    enum class StateMask {
```

Comment:
Perhaps we should do away with StateMask and just store the bool.  We could have used it for other things in RFRN but didn't.

Reply:
done